### ISSUE TITLE
[DRAFT] API code consistency

### ISSUE THREAD

**Issue by DanThrane** [SDU-eScience/UCloud#1934]
Before long, we will start seeing users who needs to integrate with the UCloud API. For a long time, our APIs have been internal. We have long been used to changing APIs as they have needed changes, in the future, we will need to take API stability and consistency more seriously.

This document covers a change our existing code and future code style to enforce a better consistency for CRUD style APIs.

The majority of APIs in UCloud/Core are CRUD style APIs for some resource of UCloud:

- Create: Creates a (sub-)resource in UCloud
- Read: Reads a single/collection of resources and optionally their associated sub-resources
  - Note: This also includes specialized read operations, such as, pagination calls
- Update: Updates a resource in UCloud
  - Note: This also includes specialized update operations, e.g., approve/reject/withdraw a grant application
- Delete: Delete a resource in UCloud

## General conventions for the Kotlin code

> Use camelCase for all class properties

> Use PascalCase for all classes

> Naming convention for `Request` and `Response` classes
> 1. Request classes should end with `Request`
> 2. Response classes should end with `Response`
> 3. The request and response classes should contain a reference to the namespace
> 4. The request and response object must reference the call name they are associated with

Valid examples:

- `JobsRetrieveRequest`
- `JobsRetrieveResponse`

Invalid examples:

- `RetrieveRequest` (no reference to namespace)
- `JobsRequest` (no reference to call)
- `JobsRetrieve` (missing suffix)
- `Retrieve` (missing suffix and namespace reference)

> Nullable fields must have a default value of null

This should be done in the constructor, this makes the Kotlin constructor consistent with the network API.

> The call name should not contain a reference to the namespace

That is, if a function will browse a list of project members the call name should be `browse` and not `browseProjectMembers`. Instead, the namespace of the call container should contain a reference to project members.

> The Kotlin file containing the call container should match the call container's name

That is, if a call container is called `Jobs` then it must be placed in `Jobs.kt`.

> Naming contention for the `CallDescriptionContainer`
>
> 1. The call container must be named after the resource it manages
> 2. The call container must always use the plural form of the resource
> 3. The call container should not use any suffix (e.g. `Descriptions`)
> 4. The call container's name should contain a the entire hierarchy   

Valid example names:

- `Jobs`
- `Avatars`
- `Projects`
- `Files`
- `ProjectMembers`

Invalid example names:

- `Job`
- `JobDescriptions`
- `FileDescriptions`
- `AvatarCallDescriptions`
- `ProjectCalls`
- `Members` (if the resource is namespaced under projects)

## General conventions for the HTTP backend

> Prefer passing the request in the body for (all methods but `GET`). Use query parameters as a fallback (`GET` only).

The request payload in an HTTP request can be passed in many ways, and we commonly see in other frameworks that this information is spread across the path, query parameters and sometimes even body as well. The UCloud convention attempts to simplify this landscape by preferring to receive the entire request via the request body. Unfortunately, the HTTP specification doesn't allow request bodies for `GET` requests (but some clients/servers do it regardless). For this reason, we choose to pass parameters purely via query parameters.

We support lists via query parameters by repeatedly passing the same parameter with different values. We will use the order they are passed in the URL.

> When accepting streaming content, pass the streamed content fully in the request body and pass additional parameters in HTTP headers.

The industry standard for accepting file uploads, and other streaming content, is to use multipart-content. This is allows an HTTP request to pass multiple values in a single HTTP request using standard HTML elements such as the `<form>`. This could, in the example of a file upload, would include the path in one part and the file in another part. Unfortunately, browsers have no real way of ordering these attributes in a specific way. This means that our servers would have to read the full request before we can fully start processing it. In many cases, we are able to quickly determine if a request is valid simply by reading the metadata. Our workaround for this is to pass all the metadata in HTTP request headers. This allows our servers to fail-fast and save both the client and server resources and time.

Fun fact: The HTTP specification allows response headers to be sent after the body, although widely unsupported.

> Use camelCase for HTTP path and query parameters but don't depend on it. The server will treat this as case-insensitive data.

UCloud generally uses camelCase for all fields in request and response payloads. We choose to consistently use this naming scheme for URLs. To improve compatibility we also treat URLs as case-insensitive.

## General conventions for the WebSocket backend

> Use a single websocket listener address for the entire namespace: `/api/$NAMESPACE`.

## Guidelines for "Create" operations

> (HTTP) Use the HTTP method `POST`.

This matches the description of the `POST` method in the HTTP specification.

> (HTTP) Path should be: `/api/$NAMESPACE`.

> All create operation should support bulk creation.

Bulk operations can be recognized by the presence of a field called "type" with a value equal to "bulk". In that case the payload must also support a field "items" containing all the resources to create.

During a bulk request the user must have permissions to create _all_ the items supplied. These will be created as a _single_ transaction. The server must respond with which elements were unable to be created.

TODO This should be implemented by creating a standard creation type which will marshall correctly from JSON while remaining backwards compatible.

> Service users should not be able to create resources on the user's behalf.

Instead a service must extend the token if they need this functionality.

> If the ID is auto-generated it should be returned as part of the response in the "id" field

## Guidelines for "Read" operations

> (HTTP) Use the HTTP method `GET`.

This matches the description of the `GET` method in the HTTP specification.

### Retrieving one or more resources ('findById' operations)

> (HTTP) Path should be `/api/$NAMESPACE/retrieve`.

We choose not to use `/api/$NAMESPACE` to make it easier to remember how to use retrieval of specific resources versus browsing resources.

> Service users should generally not be allowed to read resources on the behalf of other users.

> All of these operations must support bulk.

Bulk is supported by adding multiple query parameters to the URL.

> Additional parameters may be used to include optional data about the resource

These parameters should use the naming conventions `includeXXX` of a boolean type. These booleans should always be optional and default to `false`. The `includeXXX` properties are global for the entire request.

### Browsing a collection of resources (paginated operations)

> (HTTP) Path should be `/api/$NAMESPACE/browse`.

We choose not to use `/api/$NAMESPACE` to make it easier to remember how to use retrieval of specific resources versus browsing resources.

> Service users should generally not be allowed to read resources on the behalf of other users.

> Filters can be applied to change the output and sort order (passed via _optional_ query params).

> TODO Do we need to change the pagination API.

### Verifying information

Services quite often need to verify that information they receive is correct and verify that the requesting user is allowed to perform a certain action.

> (HTTP) Path should be `/api/$NAMESPACE/verify` or `/api/$NAMESPACE/verifyXXX`.

> (HTTP) Use the HTTP method `POST`.

> The endpoint should be privileged (`Roles.PRIVILEGED`).

> The endpoint may return additional information which may be required by the server.

> The request should accept bulk verification. Responses should be issued for the individual items.

> The request should contain an `action` field which describes the action a user is attempting to perform (if relevant).

> The request should contain an `id` field which references the resource.

> The request should contain a `username` field which references the requesting user.

> The request should contain a `project` field which references the requesting user's project.

### Searching in a collection of resources

This is only intended for when the filters by the browse operation is not enough. That is, we want provide functionality similar to a search engine.

> (HTTP) Path should be `/api/$NAMESPACE/search`.

## Guidelines for "Update" operations

This covers any type of operation which acts on a resource. This includes operations which might not directly change the output of read operations. 

> (HTTP) Use the HTTP method `POST`.

> Service users should not be allowed to perform this action on behalf of users.

> Request payload should contain a reference in the `id` field.

The only exception to this rule is if the resource itself is a singleton. For example, a compute backend might have an operation to put it in maintenance mode.

> Bulk operations should be allowed (uses similar convention to create).

## Guidelines for "Delete" operations

> (HTTP) Use the HTTP method `DELETE`

> Service users should not be allowed to perform this action on behalf of users.

> Bulk operations should be supported (same convention as create)

> The request payload should only contain a reference to the ID

> Path should be `/api/$NAMESPACE`

**Comment by DanThrane** [comment]
Design document merged into `scheduling`. Will be slowly implemented as we focus on the provider API.
